rust/src/service/data.rs

```diff
@@ -0,0 +1,191 @@
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::{
+    coordinator::{Coordinator, ProtocolEvent, RoundParameters},
+    service::handle::{SerializedSeedDict, SerializedSumDict},
+    MaskHash,
+    SeedDict,
+    SumDict,
+    SumParticipantPublicKey,
+};
+use derive_more::From;
+use sodiumoxide::crypto::box_;
+use std::{collections::HashMap, sync::Arc};
+
+/// Data that the service keeps track of.
+#[derive(From, Default)]
+pub struct Data {
+    /// Parameters of the current round. If there is no round in
+    /// progress, this is `None`.
+    pub round_parameters: Option<Arc<RoundParameters>>,
+    /// Data relevant to the current phase of the protocol. During the
+    /// update phase, this contains the sum dictionary to be sent to
+    /// the update participants for instance, while during the sum2
+    /// phase it contains the seed dictionaries.
+    pub phase_data: Option<PhaseData>,
+}
+
+/// Data held by the service in specific phases
+#[derive(From)]
+pub enum PhaseData {
+    /// Data held by the service during the sum phase
+    #[from]
+    Sum(SumData),
+
+    /// Data held by the service during the update phase
+    #[from]
+    Update(UpdateData),
+
+    /// Data held by the service during the sum2 phase
+    #[from]
+    Sum2(Sum2Data),
+
+    /// Data held by the service during the aggregation phase
+    #[from]
+    Aggregation(AggregationData),
+}
+
+impl PhaseData {
+    pub fn sum_dict(&self) -> Option<SerializedSumDict> {
```

Comment:
would be helpful to document, since it's marked public
similarly with other instances in this file